### PR TITLE
feat(backend): Refactor authz to perform SubjectAccessReview. Fixes #3513

### LINKED ISSUE #3513: Multi-User Authorization: Add support for K8s RBAC via SubjectAccessReview
As per the latest design doc: https://docs.google.com/document/d/1R9bj1uI0As6umCTZ2mv_6_tjgFshIKxkSt00QLYjNV4/edit?ts=5e4d8fbb#heading=h.5s8rbufek1ax

Add support for using K8s RBAC for authorization in Kubeflow Pipelines.
Backend endpoints will be mapped to K8s RBAC permissions.
Authorization checks will be done using SubjectAccessReview.

Related: 
https://github.com/kubeflow/pipelines/issues/1223
https://github.com/kubeflow/kubeflow/issues/4899
https://github.com/kubeflow/kubeflow/pull/4909#issuecomment-612883853

/kind feature
/area backend

/cc @IronPan @Bobgy

### PR DESCRIPTION
**Pull request by elikatsis**
**Description of your changes:**

With this PR we implement the refactoring described in #3513.

In the past, KFP used KFAM for its authorization checks, checking only if the user can `view` the corresponding namespace.
We now refactor the authorization mechanism to perform proper and fine-grained authorization via K8s RBAC and SubjectAccessReview.

Closes #3513 

/assign @Bobgy 

cc @yanniszark 

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [x] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

File: backend/src/apiserver/client/subject_access_review_fake.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
+// Copyright © 2020 Arrikto Inc.  All Rights Reserved.
```
Comment (Bobgy) [inline backend/src/apiserver/client/subject_access_review_fake.go:1]: It's okay to keep Arrikto copyright, but please add apache license header to every file.

File: backend/src/apiserver/client/subject_access_review_fake.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	authzv1 "k8s.io/api/authorization/v1"
+)
+
+type FakeSubjectAccessReviewClient struct {
+}
+
+func (FakeSubjectAccessReviewClient) Create(*authzv1.SubjectAccessReview) (*authzv1.SubjectAccessReview, error) {
+	return &authzv1.SubjectAccessReview{Status: authzv1.SubjectAccessReviewStatus{
+		Allowed:         true,
+		Denied:          false,
```
Comment (Bobgy) [inline backend/src/apiserver/client/subject_access_review_fake.go:19]: When will denied be true, what does it mean?

File: backend/src/apiserver/client/subject_access_review.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/pkg/errors"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	"k8s.io/client-go/rest"
+)
+
+func createSubjectAccessReviewClient() (v1.SubjectAccessReviewInterface, error) {
+	restConfig, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to initialize kubernetes client.")
```
Comment (Bobgy) [inline backend/src/apiserver/client/subject_access_review.go:19]: When failing with in cluster config, can we try kubeconfig, so that we can run API server locally?

File: backend/src/apiserver/client/subject_access_review.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+// CreateSubjectAccessReviewClientOrFatal creates a new SubjectAccessReview client.
+func CreateSubjectAccessReviewClientOrFatal(initConnectionTimeout time.Duration) v1.SubjectAccessReviewInterface {
+	var client v1.SubjectAccessReviewInterface
+	var err error
+	var operation = func() error {
+		client, err = createSubjectAccessReviewClient()
+		if err != nil {
+			return err
+		}
+		return nil
```
Comment (Bobgy) [inline backend/src/apiserver/client/subject_access_review.go:38]: nit: this is equivalent to 

return err

**Review by Bobgy — COMMENTED** [review]
Thank you so much @elikatsis for the contribution! I have been looking forward for this for a while!

Some early feedback first. I haven't gone through everything.

File: backend/src/apiserver/client/subject_access_review_fake.go (hunk 1 above)
Comment (elikatsis) [reply backend/src/apiserver/client/subject_access_review_fake.go:1]: Oops, I forgot this! Thanks for catching it

File: backend/src/apiserver/client/subject_access_review_fake.go (hunk 2 above)
Comment (elikatsis) [reply backend/src/apiserver/client/subject_access_review_fake.go:19]: From the [documentation](https://godoc.org/k8s.io/api/authorization/v1#SubjectAccessReviewStatus):
```go
type SubjectAccessReviewStatus struct {
    // Allowed is required. True if the action would be allowed, false otherwise.
    Allowed bool `json:"allowed" protobuf:"varint,1,opt,name=allowed"`
    // Denied is optional. True if the action would be denied, otherwise
    // false. If both allowed is false and denied is false, then the
    // authorizer has no opinion on whether to authorize the action. Denied
    // may not be true if Allowed is true.
    // +optional
    Denied bool `json:"denied,omitempty" protobuf:"varint,4,opt,name=denied"`
    ...
}
```
So I believe we only care about the `Allowed` field to decide.
Maybe if we had multiple authorization mechanisms **and** we found out that `allowed == denied == false`, then we should proceed with the next mechanism.

File: backend/src/apiserver/client/subject_access_review.go (hunk 3 above)
Comment (elikatsis) [reply backend/src/apiserver/client/subject_access_review.go:19]: If I understand correctly, this is the same with kubernetes core client initialization ([source](https://github.com/kubeflow/pipelines/blob/a8bd4682656e75b5c5cd48d75ffdf23e97bf99aa/backend/src/apiserver/client/kubernetes_core.go#L25-L36))

This means we should patch both places (and every other place) if we want to run it locally.
Would it be better to file an issue and have a distinct PR for it?

File: backend/src/apiserver/client/subject_access_review.go (hunk 4 above)
Comment (elikatsis) [reply backend/src/apiserver/client/subject_access_review.go:38]: I blindly copied this part from the kubernetes core client :smile:. I'll do this fix in both places

File: backend/src/apiserver/client/subject_access_review_fake.go (hunk 2 above)
Comment (Bobgy) [reply backend/src/apiserver/client/subject_access_review_fake.go:19]: SGTM, thanks for the info!

File: backend/src/apiserver/client/subject_access_review.go (hunk 3 above)
Comment (Bobgy) [reply backend/src/apiserver/client/subject_access_review.go:19]: Thanks, I wasn't aware of the existing convention, makes sense as separate issue and PR.

**Comment by Bobgy** [comment]
@elikatsis I've got one general question, what would be the replacement you think for KFAM for namespace discovery?
Previously, we relied on centraldashboard talking to KFAM to get that, but I believe we can no longer do so if we completely remove KFAM.

I'm thinking about listing all namespaces with a certain pipelines.kubeflow.org/xxx label and build a new selector using this info collected in backend. I wonder what do you think about it.

/cc @yanniszark

File: backend/src/apiserver/client/subject_access_review.go (hunk 3 above)
Comment (elikatsis) [reply backend/src/apiserver/client/subject_access_review.go:31]: I created an issue to track the effort: https://github.com/kubeflow/pipelines/issues/4738
So I'm marking this as resolved as well

**Comment by Bobgy** [comment]
I was kind of busy, will review tomorrow

File: backend/src/apiserver/client/subject_access_review_fake.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+		Reason:          "",
+		EvaluationError: "",
+	}}, nil
+}
+
+func (FakeSubjectAccessReviewClient) CreateContext(context.Context, *authzv1.SubjectAccessReview) (*authzv1.SubjectAccessReview, error) {
```
Comment (Bobgy) [inline backend/src/apiserver/client/subject_access_review_fake.go:37]: Have you considered making a thin interface for this client, so that we do not need to expose this method we don't use.
example: https://github.com/kubeflow/pipelines/blob/1165ed243b381ae37a72e245d685ba318d58b345/backend/src/apiserver/client/argo.go#L28-L34

**Review by Bobgy — COMMENTED** [review]
Thank you so much @elikatsis!
I just read through all the code, left some nit pickings. The high-level structure looks awesome to me. Let's follow up with some details and I think we'll be good to go.

Another general nit picking I've seen a few times is that for each resource we defined a `canAccessXXX(id)` method that takes the ID, so that we needs to first fetch this resource to know its namespace. In cases that we already fetch this resource and know its namespace, we wrote one-off boilerplate code to fetch the resource and check authorization.

I think we can simplify a little bit by defining two utility methods for each resource:
* `canAccessXxxById(id, verb)` --- fetches the resource inside and figures out namespace/name
* `canAccessXxx(namespace, name, verb)` --- we are already fetching the resource, so we can pass namespace and name directly

`canAccessXxxById` will be straightforward to implement using `canAccessXxx` and we can avoid quite some duplication there.

**Comment by Bobgy** [comment]
Other than that, can you resolve the conflicts?

File: backend/src/apiserver/client/subject_access_review_fake.go (hunk 5 above)
Comment (chensun) [reply backend/src/apiserver/client/subject_access_review_fake.go:37]: Actually where does this `CreateContext` come from? I don't see it being asked by the interface:
https://godoc.org/k8s.io/client-go/kubernetes/typed/authorization/v1#SubjectAccessReviewInterface

**Comment by maganaluis** [comment]
We are working on https://github.com/kubeflow/pipelines/issues/4197 and I just realized our changes will also have conflicts with this PR. Just for my own planning, will this be merged for the release 1.3? What else will we need to integrate with SubjectAccessReview. I know the CentralDashboard also uses KFam. 

@elikatsis @Bobgy

**Comment by Bobgy** [comment]
Yes, I think as long as @elikatsis gets through the comments, we can merge.

It should be ready with kubeflow 1.3

This change is compatible with central dashboard, profile controller and kfam, because profile controller should already give service accounts rolebindings that can aggregate ones in this PR.

**Comment by Bobgy** [comment]
It'll be ideal if we can get rid of kfam from central dashboard too, so that we can completely remove it from the manifests, but there are still some unanswered design questions.

File: manifests/kustomize/base/pipeline/cluster-scoped/view-edit-roles.yaml [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+kind: ClusterRole
+metadata:
+  labels:
+    rbac.authorization.kubeflow.org/aggregate-to-kubeflow-edit: "true"
+  name: kubeflow-pipeline-edit
+aggregationRule:
+  clusterRoleSelectors:
+  - matchLabels:
+      rbac.authorization.kubeflow.org/aggregate-to-kubeflow-pipeline-edit: "true"
+rules: []
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply manifests/kustomize/base/pipeline/cluster-scoped/view-edit-roles.yaml:11]: @Bobgy,
Great question! The answer is this essentially: https://github.com/kubernetes/kubernetes/issues/65171

TL;DR: We can't have both aggregation **and** rules in a [Cluster]Role. The rules will be ignored. This is why we need to define our rules in some [Cluster]Roles and then have _one role to find them, one role to bring them, and in itself bind them_ :smile:

File: manifests/gcp_marketplace/chart/kubeflow-pipelines/templates/pipeline.yaml [hunk 7]
```diff
@@ -273,6 +273,12 @@ rules:
       - update
       - patch
       - delete
+  - apiGroups:
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply manifests/gcp_marketplace/chart/kubeflow-pipelines/templates/pipeline.yaml]: Done

File: backend/src/apiserver/server/job_server.go [hunk 8]
```diff
@@ -88,11 +89,37 @@ func (s *JobServer) CreateJob(ctx context.Context, request *api.CreateJobRequest
 	if err != nil {
 		return nil, util.Wrap(err, "Validate create job request failed.")
 	}
-	err = CanAccessExperimentInResourceReferences(s.resourceManager, ctx, request.Job.ResourceReferences)
 	if err != nil {
 		return nil, util.Wrap(err, "Failed to authorize the request.")
 	}
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/server/job_server.go]: Yes! Thanks for catching it

**Comment by elikatsis** [comment]
@Bobgy 
I rebased and force-pushed. The minor nits (gcp marketplace manifests and forgotten double error check) are included in the "old" commits.
For the comments about
* centralized check for multi-user shared mode
* use `canAccessXXX()` instead of `isAuthorized()`
* enhancement on error assertion wrt unauthorized calls
I have added two fixup commits. I will squash them if they look good to you.

As far as having general `canAccessXXX()` and `canAccessXXXById()` functions, I think we don't need them. This is covered by each server's methods (see first `fixup!` commit). Implementing them in a different place and have the server use them looks like it adds complexity instead of simplifying things.

File: backend/src/apiserver/server/experiment_server.go [hunk 9]
```diff
@@ -137,7 +147,16 @@ func (s *ExperimentServer) ListExperiment(ctx context.Context, request *api.List
...
+		resourceAttributes := &authorizationv1.ResourceAttributes{
+			Namespace:   namespace,
+			Verb:        common.RbacResourceVerbList,
+			Group:       common.RbacPipelinesGroup,
+			Version:     common.RbacPipelinesVersion,
+			Resource:    common.RbacResourceTypeExperiments,
+			Subresource: "",
+			Name:        "",
+		}
+		err = isAuthorized(s.resourceManager, ctx, resourceAttributes)
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/server/experiment_server.go]: Done

File: backend/src/apiserver/server/auth_server_test.go [hunk 10]
```diff
@@ -94,7 +94,7 @@ func TestAuthorizeRequest_Unauthorized(t *testing.T) {
 
 	_, err := authServer.Authorize(ctx, request)
 	assert.Error(t, err)
-	assert.EqualError(t, err, "Failed to authorize the request: Failed to authorize namespace: BadRequestError: Unauthorized access for user@google.com to namespace ns1: Unauthorized access for user@google.com to namespace ns1")
+	assert.Contains(t, err.Error(), "Unauthorized access")
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/server/auth_server_test.go]: Done (2nd fixup commit). There's quite some refactoring there so please tell me what do you think :)

File: backend/src/apiserver/client/subject_access_review.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	var operation = func() error {
+		client, err = createSubjectAccessReviewClient()
+		return err
+	}
+	b := backoff.NewExponentialBackOff()
+	b.MaxElapsedTime = initConnectionTimeout
+	err = backoff.Retry(operation, b)
+
+	if err != nil {
+		glog.Fatalf("Failed to create pod client. Error: %v", err)
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/client/subject_access_review.go:54]: Fixed

File: backend/src/apiserver/client/subject_access_review_fake.go (hunk 5 above)
Comment (elikatsis) [reply backend/src/apiserver/client/subject_access_review_fake.go]: We are using client-go v0.17.9: https://pkg.go.dev/k8s.io/client-go@v0.17.9/kubernetes/typed/authorization/v1#SelfSubjectAccessReviewInterface

I wrote a custom `SubjectAccessReviewInterface` which we can remove it in the future, when we upgrade `client-go`.

File: manifests/kustomize/base/pipeline/cluster-scoped/view-edit-roles.yaml [hunk 12]
```diff
@@ -0,0 +1,108 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply manifests/kustomize/base/pipeline/cluster-scoped/view-edit-roles.yaml]: Sure, we will get the manifests repo updated accordingly as soon as we get this merged.
I created the directory you mentioned and moved the role declarations there.

File: backend/src/apiserver/server/util.go [hunk 13]
```diff
@@ -289,110 +290,49 @@ func getUserIdentityFromHeader(userIdentityHeader, prefix string) (string, error
...
-func CanAccessExperiment(resourceManager *resource.ResourceManager, ctx context.Context, experimentID string) error {
+// isAuthorized verifies whether the user identity, which is contained in the context object,
+// can perform some action (verb) on a resource (resourceType/resourceName) living in the
+// target namespace. If the returned error is nil, the authorization passes. Otherwise,
+// authorization fails with a non-nil error.
+func isAuthorized(resourceManager *resource.ResourceManager, ctx context.Context, resourceAttributes *authorizationv1.ResourceAttributes) error {
 	if common.IsMultiUserMode() == false {
 		// Skip authz if not multi-user mode.
 		return nil
 	}
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply backend/src/apiserver/server/util.go:316]: I think I covered this with the `fixup!` commits. PTAL

File: backend/src/apiserver/server/util.go [hunk 14]
```diff
@@ -289,110 +290,55 @@ func getUserIdentityFromHeader(userIdentityHeader, prefix string) (string, error
...
-
-	if isAuthorized == false {
-		glog.Infof("Unauthorized access for %s to namespace %s", userIdentity, namespace)
-		return util.NewBadRequestError(errors.New("Unauthorized access for "+userIdentity+" to namespace "+namespace), "Unauthorized access for "+userIdentity+" to namespace "+namespace)
+		glog.Info(err.Error())
+		return err
 	}
 
-	glog.Infof("Authorized user %s in namespace %s", userIdentity, namespace)
+	glog.Infof("Authorized user '%s': %+v", userIdentity, resourceAttributes)
```
Comment (Bobgy) [inline backend/src/apiserver/server/util.go:342]: These authorization logs will be very common, every request comes with it.
Shall we change log level to `trace` or `debug`?

For most of the cases, users should depend on error message for debugging, rather than api server logs.

File: manifests/kustomize/base/pipeline/cluster-scoped/view-edit-roles.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (Bobgy) [reply manifests/kustomize/base/pipeline/cluster-scoped/view-edit-roles.yaml]: Thank you! That would be great

File: manifests/kustomize/base/pipeline/cluster-scoped/view-edit-roles.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Bobgy) [reply manifests/kustomize/base/pipeline/cluster-scoped/view-edit-roles.yaml]: Thanks for the explanation! Good to know. : )

File: manifests/kustomize/base/pipeline/multi-user/view-edit-roles.yaml [hunk 15]
```diff
@@ -0,0 +1,108 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (Bobgy) [inline manifests/kustomize/base/pipeline/multi-user/view-edit-roles.yaml:1]: @elikatsis Can you add an inline comment with short explanation? I think upcomers will be confused too, so it'll be great as a comment.

**Review by Bobgy — COMMENTED** [review]
I've only got two comments left. Everything else looks great to me!
Thank you so much @elikatsis!

/lgtm

File: backend/src/apiserver/server/util.go (hunk 14 above)
Comment (elikatsis) [reply backend/src/apiserver/server/util.go:342]: I can only see `Info`, `Warning`, `Error`, and `Fatal` levels. Can you help me how can I do `debug` or sth different?
Here's what I found in the repo, not sure if it helps: https://github.com/golang/glog/blob/23def4e6c14b4da8ac2ed8007337bc5eb5007998/glog.go#L91-L115

**Comment by elikatsis** [comment]
@Bobgy I added the comment and squashed the fixups. If you can help me selecting the logging level, we'll be golden!

File: backend/src/apiserver/server/util.go (hunk 14 above)
Comment (Bobgy) [reply backend/src/apiserver/server/util.go:342]: OK, sorry I thought there were more. Then let's go with info for now.

**Comment by Bobgy** [comment]
Do you need @yanniszark 's review?
Everything good on my side

**Comment by elikatsis** [comment]
@Bobgy we have already reviewed this internally, so we should be good to go. Thank you very much for your comments! :smiley:

**Comment by Bobgy** [comment]
/lgtm
/approve
Cool, I am feeling excited about the huge potential this brings. Thank you again for this great contribution!
